vulnerability-scan/index.js

```diff
@@ -0,0 +1,382 @@
+const core = require("@actions/core");
+const fetch = require("node-fetch");
+async function run() {
+    const jiraUsername = core.getInput("jira_username", { required: true });
+    const jiraToken = core.getInput("jira_token", { required: true });
+    const quayToken = core.getInput("quay_token", { required: true });
+    const quayTag = core.getInput("quay_tag", { required: true });
+    const quayContainersInput = core.getInput("quay_containers", { required: true });
+    const quayContainers = quayContainersInput.split(',').map(container => container.trim());
+
+    const authToken =
+        "Basic " +
+        Buffer.from(jiraUsername + ":" + jiraToken).toString("base64");
+
+    for (const quayContainer of quayContainers) {
```

Comment:
I think we should restructure this approach. Instead of going by container by container to do all of this, could we instead:

Loop over the containers and store all of their vulnerabilities, maybe something like this:

```javascript
const vulnerabilities = {};
for (const quayContainer of quayContainers) {
  const containerVulns = await getVulnerabilities(
    quayToken,
    quayContainer,
    tag);  // move digest retrieval inside getVulns

  // filterVulns handles severity, fixed-by, etc
  for (const vuln of filterVulns(containerVulns)) {
    const thisVuln = vulnerabilities[vuln.CVEId] || (vulnerabilities[vuln.CVEId] = {});

    if (thisVuln.CVEId === undefined) {
      thisVuln.containers = [quayContainer];
      thisVuln.CVEId = vuln.CVEId;
      // etc
    } else {
      thisVuln.containers.push(quayContainer);
    }
  }
}
```

After that we can then handle the JIRA tickets:

```javascript
for (const cveId of vulnerabilities) {
  const vuln = vulnerabilities[cveId];
  const search = await searchJiraIssue(authToken, cveId);

  switch (search.issues.length) {
    case 0:
      await createJiraIssue(authToken, vuln);
      break;
    case 1:
      const ticket = search.issues[0];
      await updateJiraIssue(authToken, ticket, vuln);
      break
    default:
      throw new Error(`Expected 0 or 1 Jira issues but found ${search.issues.length}`);
  }
}
```

Reply:
> `await searchJiraIssue(authToken, cveId);`

If we bunch up all of the vulnerabilities we could look for all of our tickets at the same time, but I didn't want to depart too far from what was here.

Reply:
reworked it a bit to be more like what you were saying. Hopefully it clears it up a bit.